apps/emqx_durable_storage/src/emqx_ds_cache_worker.erl

```diff
@@ -0,0 +1,226 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_ds_cache_worker).
+
+-behaviour(gen_server).
+
+%% API
+-export([
+    start_link/4,
+
+    get_table/2
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+-include_lib("snabbkaffe/include/trace.hrl").
+-include("emqx_ds_cache.hrl").
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-define(APP, emqx_durable_storage).
+-define(via(REC), {via, gproc, {n, l, REC}}).
+-define(cache_worker, ?MODULE).
+
+-record(?cache_worker, {db, stream}).
+
+-type state() :: #{
+    db := emqx_ds:db(),
+    table := ets:tid(),
+    iterator := ?EOS_MARKER | emqx_ds:iterator(),
+    seqno := undefined | seqno()
+}.
+
+%% call/cast/info records
+-record(get_table, {}).
+-record(pull, {}).
+
+%%================================================================================
+%% API
+%%================================================================================
+
+-spec start_link(emqx_ds:db(), emqx_ds:stream(), emqx_ds:topic_filter(), emqx_ds:time()) ->
+    supervisor:startchild_ret().
+start_link(DB, Stream, TopicFilter, StartTime) ->
+    gen_server:start_link(
+        ?via(#?cache_worker{db = DB, stream = Stream}),
+        ?MODULE,
+        {DB, Stream, TopicFilter, StartTime},
```

Comment:
Why do we need to supply topic filter and start time to the cache worker? It seems inconsistent with the process ID, that only allows to have one cache worker for the DB and the stream.
If different different topic filters happen to map to the same stream, does it mean they will compete for the process registration?

Reply:
The current API to create iterators require the topic filter to be provided, so I don't think there's currently a way around it?

Reply:
Also, we've discussed that the first version of the cache could use statically configured topic filters, and later make it automatically track streams based on usage or other heuristics.  So the future version might not need to specify the topic filter and use `#`.

Reply:
I've just found a fundamental problem with the cache right now: it is currently just relying on seqnos to detect gaps, without checking that the message topic actually matches the iterator topic filter...  It'll require some rethinking.  :see_no_evil: